crates/ruff_server/src/server/api/requests/execute_command.rs

```diff
@@ -0,0 +1,213 @@
+                    edit_tracker
+                        .add_edits_for_document(uri, version, edits)
+                        .with_failure_code(ErrorCode::InternalError)?;
+                }
+                Command::Format => {
+                    let response = super::format::format_document(&snapshot)?;
+                    if let Some(edits) = response {
+                        edit_tracker
+                            .add_edits_for_document(uri, version, edits)
+                            .with_failure_code(ErrorCode::InternalError)?;
+                    }
+                }
+                Command::OrganizeImports => {
+                    let edits = super::code_action_resolve::organize_imports_edit(
+                        snapshot.document(),
+                        &snapshot.configuration().linter,
+                        snapshot.encoding(),
+                    )
+                    .with_failure_code(ErrorCode::InternalError)?;
+                    edit_tracker
+                        .add_edits_for_document(uri, version, edits)
+                        .with_failure_code(ErrorCode::InternalError)?;
+                }
+            }
+        }
+
+        if !edit_tracker.is_empty() {
+            apply_edit(
+                requester,
+                command.label(),
+                edit_tracker.into_workspace_edit(),
+            )
+            .with_failure_code(ErrorCode::InternalError)?;
+        }
+
+        Ok(None)
+    }
+}
+
+impl Command {
+    fn label(&self) -> &str {
+        match self {
+            Self::FixAll => "Fix all auto-fixable problems",
+            Self::Format => "Format document",
+            Self::OrganizeImports => "Format imports",
+        }
+    }
+}
+
+impl FromStr for Command {
+    type Err = anyhow::Error;
+
+    fn from_str(name: &str) -> Result<Self, Self::Err> {
+        Ok(match name {
+            "ruff.applyAutofix" => Self::FixAll,
+            "ruff.applyFormat" => Self::Format,
+            "ruff.applyOrganizeImports" => Self::OrganizeImports,
+            _ => return Err(anyhow::anyhow!("Invalid command `{name}`")),
+        })
+    }
+}
+
+impl CommandEditTracker {
+    fn new(document_changes_supported: bool) -> Self {
+        if document_changes_supported {
+            Self::DocumentChanges(Vec::default())
+        } else {
+            Self::Changes(HashMap::default())
+        }
+    }
+
+    fn add_edits_for_document(
+        &mut self,
+        uri: types::Url,
+        version: DocumentVersion,
+        edits: Vec<types::TextEdit>,
+    ) -> crate::Result<()> {
+        match self {
+            Self::DocumentChanges(document_edits) => {
+                if document_edits
+                    .iter()
+                    .any(|document| document.text_document.uri == uri)
+                {
+                    return Err(anyhow::anyhow!("Attempted to add edits for a document that was already edited - did you pass in the same document twice?"));
```

Comment:
I naively would've assumed that this would _replace_ or _augment_ the existing edits, rather than erroring. I might suggest making it a replacement (`set_edits_for_document`)?

Reply:
It seems like one of the goals of `document_changes` is that it's _ok_ to repeat a URI as long as the version differs? So perhaps this check should be on URI and version?

Reply:
Still, I don't think it makes sense to apply document edits to the same document at different points in time simultaneously.